File: lib/vac_generator.rb
# frozen_string_literal: true

require "lib/command_set"

class VACGenerator
  attr_reader :command_set

  def initialize(command_set)
    @command_set = command_set
    @builder     = Nokogiri::XML::Builder.new { generate_xml(_1) }
  end

  def to_s() @builder.to_xml(indent: 4) end

  private

  def generate_xml(xml)
    xml.profile do
      command_set.each_command { |command| output_command xml, command }

      xml.set key:  "NONE", key1: "NONE", vackey1: "NONE", vackey2: "NONE",
              ptamode: "default", off: "NONE", off1: "NONE", off2: "NONE",
              off3: "NONE"
    end
  end

  def output_command(xml, command)
    return if command.virtual?

    xml.command(xml_attributes(command)) do
      command.macro.each do |keystroke|
        if extended?(keystroke)
          xml.key extended:  keystroke,
                  extended2: "NONE",
                  qual:      "NONE",
                  pause:     60,
                  repeat:    1,
                  duration:  60
        else
          xml.key value:     keystroke,
                  extended:  "NONE",
                  extended2: "NONE",
                  qual:      "NONE",
                  pause:     60,
                  repeat:    1,
                  duration:  60
        end
      end
    end
  end

  def extended?(keystroke)
    keystroke.match? CommandSet::EXTENDED_KEYSTROKES_RX
  end

  def xml_attributes(command)
    attrs = {name: command.name}
    command.full_phrases.each_with_index do |phrase, index|
      key               = index.zero? ? "phrase" : "phrase#{index}"
      attrs[key.to_sym] = phrase
    end

    return attrs
  end
end

File: Sources/VoiceRecognitionProfileGenerator/VACGenerator.swift
import Foundation

class VACGenerator: Generator {
  private let commands: CommandSet

  var defaultKeystrokeAttributes = [
    "qual": "NONE",
    "pause": "60",
    "repeat": "1",
    "duration": "60"
  ]

  required init(commands: CommandSet) {
    self.commands = commands
  }

  func generate() throws -> String {
    let root = XMLElement(name: "profile")
    let xml = XMLDocument(rootElement: root)

    try commands.each { command in
      guard commands.isReal(command) else { return }
      try root.addChild(xmlElement(for: command))
    }

    root.addChild(
      XMLElement.make(
        "set",
        attributes: [
          "key": "NONE",
          "key1": "NONE",
          "vackey1": "NONE",
          "vackey2": "NONE",
          "ptamode": "default",
          "off": "NONE",
          "off1": "NONE",
          "off2": "NONE",
          "off3": "NONE"
        ]
      )
    )

    return xml.xmlString(options: .nodePrettyPrint)
  }

  private func xmlElement(for command: Command) throws -> XMLElement {
    let element = XMLElement(name: "command")
    element.setAttribute(name: "name", value: command.name)

    for (index, phrase) in command.fullPhrases.enumerated() {
      let name = (index == 0 ? "phrase" : "phrase\(index)")
      element.setAttribute(name: name, value: phrase)
    }

    for keystrokes in command.macro {
      guard let keystroke = keystrokes.first, keystrokes.count == 1 else {
        throw GeneratorErrors.chordingUnsupported
      }
      try element.addChild(xmlElement(for: keystroke))
    }

    return element
  }

  private func xmlElement(for keystroke: Keystroke) throws -> XMLElement {
    let element = XMLElement(name: "key")
    element.addAttributes(defaultKeystrokeAttributes)

    switch keystroke {
      case let .character(char):
        element.addAttributes([
          "value": String(char),
          "extended": "NONE",
          "extended2": "NONE"
        ])
      case let .function(num):
        element.addAttributes([
          "extended": "F\(num)",
          "extended2": "NONE"
        ])
      default: throw GeneratorErrors.unsupportedKeystroke(keystroke)
    }

    return element
  }
}
